### PR TITLE
608 switch lcsh lcnaf source

### PR DESCRIPTION
**Pull request by sharpattack**

### CODE REVIEW

File: app/controllers/custom_authorities_controller.rb [hunk 1]
```diff
@@ -1,20 +1,24 @@
 class CustomAuthoritiesController < ApplicationController
   def query_mesh
-    terms = []
+    base_sparql_mesh_uri ="https://id.nlm.nih.gov/mesh/sparql?format=JSON&limit=10&inference=true&query=PREFIX%20rdfs%3A%20%3Chttp%3A%2F%2Fwww.w3.org%2F2000%2F01%2Frdf-schema%23%3E%0D%0APREFIX%20meshv%3A%20%3Chttp%3A%2F%2Fid.nlm.nih.gov%2Fmesh%2Fvocab%23%3E%0D%0APREFIX%20mesh2018%3A%20%3Chttp%3A%2F%2Fid.nlm.nih.gov%2Fmesh%3E%0D%0A%0D%0ASELECT%20%3Fd%20%3FdName%0D%0AFROM%20%3Chttp%3A%2F%2Fid.nlm.nih.gov%2Fmesh%3E%0D%0AWHERE%20%7B%0D%0A%20%20%3Fd%20a%20meshv%3ADescriptor%20.%0D%0A%20%20%3Fd%20rdfs%3Alabel%20%3FdName%0D%0A%20%20FILTER(REGEX(%3FdName%2C%27"
```
Comment (phebal) [inline app/controllers/custom_authorities_controller.rb]: Wow!

File: app/controllers/authorities_controller.rb [hunk 2]
```diff
@@ -4,11 +4,31 @@
...
   def query
     s = params.fetch('q', '')
-    hits = if ['location', 'based_near'].include?(params[:term])
+    term = params[:term]
+    hits = if ['location', 'based_near'].include?(term)
       GeoNamesResource.find_location(s)
+    elsif ['lcsh', 'subject_geographic'].include?(term)
+      get_results(term, s)
     else
       LocalAuthority.entries_by_term(params[:model], params[:term], s) rescue []
```
Comment (phebal) [inline app/controllers/authorities_controller.rb]: What's this for now?

File: app/controllers/authorities_controller.rb [hunk 3]
```diff
@@ -4,11 +4,31 @@
...
+    # the /suggest path limits results to 10
+    lcsh_base_uri = "http://id.loc.gov/authorities/subjects/suggest/?q="
+    # can increase limit to max 20 with '&rows=20'
+    fast_geo_uri = "http://fast.oclc.org/searchfast/fastsuggest?fl=suggest51&wt=json&query="
+    if subject == 'lcsh'
+      result = JSON.parse(HTTParty.get(
+        lcsh_base_uri + "#{search.gsub(/\s/,'*')}*"
+      ))[1]
+    elsif subject == 'subject_geographic'
+      result = HTTParty.get(fast_geo_uri + search)["response"]["docs"]
```
Comment (phebal) [inline app/controllers/authorities_controller.rb]: Is this guaranteed to be an array?

File: app/controllers/custom_authorities_controller.rb [hunk 4]
```diff
@@ -1,20 +1,24 @@
...
+    end_sparql_mesh_uri = "%27%2C%20%27i%27))%20%0D%0A%7D%20%0D%0AORDER%20BY%20%3Fd%20%0D%0A"
     if params[:q].present?
-      authority = Qa::Authorities::Mesh.new
-      terms = authority.search(params[:q].to_s.downcase)
+      hits = JSON.parse(HTTParty.get(
+        base_sparql_mesh_uri + params[:q] + end_sparql_mesh_uri
+      ))["results"]["bindings"]
```
Comment (phebal) [inline app/controllers/custom_authorities_controller.rb]: Same as above, is this guaranteed to be an array?

File: spec/controllers/authorities_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+      subject { JSON.parse(get(:query, q: 'ethiopia', term: 'based_near').body) }
+      
+      it 'returns values from GeoNames' do
+        expect(subject.count).to eq(10)
+        subject.each{ |s|
+          expect(s["value"].downcase).to include('ethiopia')
+        }
+      end
+    end
+  end
```
Comment (phebal) [inline spec/controllers/authorities_controller_spec.rb]: Missing tests for queries not returning results and empty queries.

File: spec/controllers/custom_authorities_controller_spec.rb [hunk 6]
```diff
@@ -195,82 +195,77 @@
...
-      Qa::SubjectMeshTerm.create(term_id: 'A', term: 'Aaa', term_lower: 'abc')
-      Qa::SubjectMeshTerm.create(term_id: 'B', term: 'Bbb', term_lower: 'bbb')
-      Qa::SubjectMeshTerm.create(term_id: 'C', term: 'BBc', term_lower: 'bbc')
-    end
-
+  describe '#query_mesh', :vcr do
     context 'invalid query' do
       describe 'no params passed' do
-        subject { get :query_mesh}
+        subject { get :query_mesh }
```
Comment (phebal) [inline spec/controllers/custom_authorities_controller_spec.rb]: thank you ;)

File: spec/rails_helper.rb [hunk 7]
```diff
@@ -12,14 +12,27 @@
 include Warden::Test::Helpers
 Warden.test_mode!
 
+options = {
+  js_errors: false,
```
Comment (phebal) [inline spec/rails_helper.rb]: Do you know why we have this on?

File: app/controllers/custom_authorities_controller.rb (hunk 1 above)
Comment (sharpattack) [reply app/controllers/custom_authorities_controller.rb]: i know, it's the format for fitting a sparql query in a url.

File: app/controllers/authorities_controller.rb (hunk 3 above)
Comment (sharpattack) [reply app/controllers/authorities_controller.rb]: you mean in case of one result? i can test for it.

File: spec/rails_helper.rb (hunk 7 above)
Comment (sharpattack) [reply spec/rails_helper.rb]: i do not, i just copied over the same vcr configurations for the website.

File: spec/controllers/authorities_controller_spec.rb (hunk 5 above)
Comment (sharpattack) [reply spec/controllers/authorities_controller_spec.rb]: we require at least two characters before making the request, should i still test for empty queries?

File: app/controllers/authorities_controller.rb [hunk 8]
```diff
@@ -20,13 +20,20 @@ def get_results(subject, search)
     lcsh_base_uri = "http://id.loc.gov/authorities/subjects/suggest/?q="
     # can increase limit to max 20 with '&rows=20'
     fast_geo_uri = "http://fast.oclc.org/searchfast/fastsuggest?fl=suggest51&wt=json&query="
+    
     if subject == 'lcsh'
       result = JSON.parse(HTTParty.get(
         lcsh_base_uri + "#{search.gsub(/\s/,'*')}*"
-      ))[1]
+      ))
+      result = result[1] ? result[1] : ["Error Searching"]
```
Comment (phebal) [inline app/controllers/authorities_controller.rb]: So `result` is guaranteed to be an array?

File: app/controllers/authorities_controller.rb [hunk 9]
```diff
@@ -20,13 +20,20 @@ def get_results(subject, search)
...
     elsif subject == 'subject_geographic'
-      result = HTTParty.get(fast_geo_uri + search)["response"]["docs"]
-      result.map!{ |x| x["suggest51"] }
+      result = HTTParty.get(fast_geo_uri + search)
+      if result["response"]["docs"]
```
Comment (phebal) [inline app/controllers/authorities_controller.rb]: you can just do `if result = result["response"]["docs"]`

File: app/controllers/custom_authorities_controller.rb [hunk 10]
```diff
@@ -5,8 +5,13 @@ def query_mesh
     if params[:q].present?
       hits = JSON.parse(HTTParty.get(
         base_sparql_mesh_uri + params[:q] + end_sparql_mesh_uri
-      ))["results"]["bindings"]
-      hits.map!{|x| x["dName"]["value"]}
+      ))
+      if hits["results"]["bindings"]
```
Comment (phebal) [inline app/controllers/custom_authorities_controller.rb]: Same here `if hits = hits["results"]["bindings"]`

File: app/controllers/custom_authorities_controller.rb [hunk 11]
```diff
@@ -16,8 +21,10 @@ def lcsh_names
     if params[:q].present?
       hits = JSON.parse(HTTParty.get(
         lcnaf_base_uri + "*#{params[:q].gsub(/\s/,'*')}*"
-      ))[1]
+      ))
+      hits = hits[1] ? hits[1] : ["Error Searching"]
```
Comment (phebal) [inline app/controllers/custom_authorities_controller.rb]: Guaranteed to be an array?

File: spec/controllers/authorities_controller_spec.rb (hunk 5 above)
Comment (phebal) [reply spec/controllers/authorities_controller_spec.rb]: Nah

**Comment by phebal** [comment]
I've tested the resque service fix on staging and in production.

File: app/controllers/authorities_controller.rb (hunk 8 above)
Comment (sharpattack) [reply app/controllers/authorities_controller.rb]: it is now, right? it should default to `["Error Searching"]`, no?

File: app/controllers/authorities_controller.rb (hunk 9 above)
Comment (sharpattack) [reply app/controllers/authorities_controller.rb]: ah, i didn't think of this. should i refactor?

Comment (phebal) [reply app/controllers/authorities_controller.rb]: it's up to you
